nni/runtime/log.py

```diff
@@ -7,14 +7,149 @@
 import sys
 from datetime import datetime
 from io import TextIOBase
-from logging import FileHandler, Formatter, Handler, StreamHandler
+from logging import FileHandler, Formatter, Handler, LogRecord, Logger, StreamHandler
 from pathlib import Path
+import string
 from typing import Optional
 
 import colorama
 
 from .env_vars import dispatcher_env_vars, trial_env_vars
 
+_global_logging_enabled: bool = False
+
+_root_logger: Logger = logging.getLogger()
+logging.getLogger('nni').setLevel(logging.INFO)
+
+_handlers: dict[str, Handler] = {}
+
+colorama.init()
+_colorful: dict[str, str] = colorama.Fore.__dict__
+_colorless: dict[str, str] = {key: '' for key in _colorful.keys()}
+
+def start_stdout_logging() -> None:
+    """
+    Register the stdout handler.
+
+    This function should be invoked on importing nni.
+
+    It is safe to call it multiple times.
+    """
+    if '_stdout_' in _handlers:
+        return
+
+    handler = StreamHandler(sys.stdout)
+    handler.addFilter(_nni_filter)
+    handler.setFormatter(_StdoutFormatter())
+
+    _handlers['_stdout_'] = handler
+    _root_logger.addHandler(handler)
+
+def silence_stdout() -> None:
+    """
+    Stop NNI from printing to stdout.
+
+    By default NNI prints log messages of ``INFO`` or higher level to console.
+    Use this function if you want a clean stdout, or if you want to handle logs yourself.
+    """
+    handler = _handlers.pop('_stdout_', None)
+    if handler is not None:
+        _root_logger.removeHandler(handler)
+
+def start_experiment_logging(experiment_id: str, log_file: Path, level: str) -> None:
+    """
+    Register the log file handler for an experiment's ``experiment.log``.
+
+    This function should be invoked on starting experiment.
+    We don't want to create the experiment folder if the user does not launch it.
+
+    If there are multiple experiments running concurrently,
+    log messages will be written to all running experiments' log file.
+
+    It is safe to call it multiple times.
+    """
+    if experiment_id in _handlers:
+        return
```

Comment:
It "should" not happen but we want robustness for edge cases.
For example if a user tries to create experiment in interactive Python and `start()` raises exception, it may leave uncleaned handlers.